packages/core/src/dialects/abstract/query-generator-internal.ts

```diff
@@ -305,11 +315,99 @@ Only named replacements (:name) are allowed in literal() because we cannot guara
   }
 
   /**
-   * Returns an SQL fragment for adding result constraints.
+   * Returns an SQL fragment for adding LIMIT and/or OFFSET to a query.
    *
    * @param _options
    */
   addLimitAndOffset(_options: AddLimitOffsetOptions): string {
     throw new Error(`addLimitAndOffset has not been implemented in ${this.dialect.name}.`);
   }
+
+  /**
+   * Returns a SQL fragment for a WHERE with a LIMIT/OFFSET using a subquery.
+   *
+   * @param tableName
+   * @param options
+   */
+  addWhereSelectLimitOffset(tableName: TableNameOrModel, options: AddWhereSelectLimitOffset): string {
+    const model = isModelStatic(tableName) ? tableName : options.model;
+    const whereOptions = { ...options, model };
+    const whereQuery = whereOptions.where ? this.queryGenerator.whereQuery(whereOptions.where, whereOptions) : '';
+    if (whereOptions.limit && this.dialect.supports.modelWithLimit) {
+      if (!model) {
+        throw new Error('Cannot use LIMIT without a model.');
+      }
+
+      const pks = Object.values(model.primaryKeys).map(key => this.queryGenerator.quoteIdentifier(key.columnName)).join(', ');
+      const primaryKeys = Object.values(model.primaryKeys).length > 1 ? `(${pks})` : pks;
+
+      return joinSQLFragments([
+        `WHERE ${primaryKeys} IN (`,
+        `SELECT ${pks} FROM ${this.queryGenerator.quoteTable(tableName)}`,
+        whereQuery,
+        `ORDER BY ${pks}`,
+        this.addLimitAndOffset(whereOptions),
+        ')',
+      ]);
+    }
+
+    return joinSQLFragments([
+      whereQuery,
+      this.addLimitAndOffset(whereOptions),
+    ]);
+  }
+
+  /**
+   * Creates a function that can be used to collect bind parameters.
+   *
+   * @param bind A mutable object to which bind parameters will be added.
+   */
+  bindParam(bind: Record<string, unknown>): (value: unknown) => string {
+    let i = 0;
+
+    return (value: unknown): string => {
+      const bindName = `sequelize_${++i}`;
+
+      bind[bindName] = value;
+
+      return `$${bindName}`;
+    };
+  }
+
+  /**
+   * Returns the SQL fragment to handle returning the attributes from an insert/update query.
+   *
+   * @param options         An object with options.
+   * @param modelAttributes A map with the model attributes.
+  */
+  getReturnFields(options: GetReturnFieldsOptions, modelAttributes: Map<string, AttributeOptions>): string[] {
+    const returnFields: string[] = [];
+
+    if (Array.isArray(options.returning)) {
+      returnFields.push(...options.returning.map(field => {
+        if (typeof field === 'string') {
+          return this.queryGenerator.quoteIdentifier(field);
+        } else if (field instanceof Literal) {
```

Comment:
if we accept literal, we should likely just accept `BaseSqlExpression`, as `literal` can include all other types

This means that (when available), `getReturnFields` should pass down things Format options `replacements` and `model` to `formatSqlExpression`

Reply:
Done, I had to add a private property to the `BaseSqlExpression` class otherwise it would allow any value.